### PR TITLE
set deadline correctly fix #4

### LINKED ISSUE #4: `Set*Deadline` doesn't work well when called with zero value
Use int64 to represent deadline is buggy, see https://play.golang.org/p/kjQlmotLmp

Relevant code:
https://github.com/getlantern/idletiming/blob/master/idletiming_conn.go#L256
https://github.com/getlantern/idletiming/blob/master/idletiming_conn.go#L268

### PR DESCRIPTION
**Pull request by fffw**

### CODE REVIEW

File: idletiming_conn.go [hunk 1]
```diff
@@ -69,11 +68,11 @@ func Conn(conn net.Conn, idleTimeout time.Duration, onIdle func()) *IdleTimingCo
 // IdleTimingConn is a net.Conn that wraps another net.Conn and that times out
 // if idle for more than idleTimeout.
 type IdleTimingConn struct {
-	// Keep them at the top to make sure 64-bit alignment, see
+	// Keep it at the top to make sure 64-bit alignment, see
 	// https://golang.org/pkg/sync/atomic/#pkg-note-BUG
-	readDeadline     int64
-	writeDeadline    int64
 	lastActivityTime int64
+	readDeadline     guardedTime
```
Comment (fffw) [inline idletiming_conn.go]: We can convert between epoch and zero value when `Set*Deadline`, but that makes the code even harder to read. Consider `idleTimeout` is usually in seconds, the performance gain by using atomic operations is neglectable.

File: idletiming_conn.go [hunk 2]
```diff
@@ -302,3 +301,21 @@ func isTimeout(err error) bool {
...
 	return false
 }
+
+type guardedTime struct {
+	sync.RWMutex
+	t time.Time
+}
+
+func (g *guardedTime) Get() time.Time {
+	g.RLock()
```
Comment (fffw) [inline idletiming_conn.go]: `defer` is kinda expensive. Safe to remove it here.

**Comment by fffw** [comment]
This PR should fix https://github.com/getlantern/lantern/issues/5023 . @oxtoacart does it look good to you?

**Comment by oxtoacart** [comment]
LGTM!  Will you use `gost get` to pull this into the lantern-pro repo?

**Comment by fffw** [comment]
Thanks yeah will pull and verify

**Pull request merged**